overlord/ifacestate/apparmorprompting/common/common.go

```diff
@@ -0,0 +1,328 @@
+	}
+	if p&apparmor.MayDeletePermission != 0 {
+		perms = append(perms, PermissionDelete)
+	}
+	if p&apparmor.MayOpenPermission != 0 {
+		perms = append(perms, PermissionOpen)
+	}
+	if p&apparmor.MayRenamePermission != 0 {
+		perms = append(perms, PermissionRename)
+	}
+	if p&apparmor.MaySetAttrPermission != 0 {
+		perms = append(perms, PermissionSetAttr)
+	}
+	if p&apparmor.MayGetAttrPermission != 0 {
+		perms = append(perms, PermissionGetAttr)
+	}
+	if p&apparmor.MaySetCredentialPermission != 0 {
+		perms = append(perms, PermissionSetCred)
+	}
+	if p&apparmor.MayGetCredentialPermission != 0 {
+		perms = append(perms, PermissionGetCred)
+	}
+	if p&apparmor.MayChangeModePermission != 0 {
+		perms = append(perms, PermissionChangeMode)
+	}
+	if p&apparmor.MayChangeOwnerPermission != 0 {
+		perms = append(perms, PermissionChangeOwner)
+	}
+	if p&apparmor.MayChangeGroupPermission != 0 {
+		perms = append(perms, PermissionChangeGroup)
+	}
+	if p&apparmor.MayLockPermission != 0 {
+		perms = append(perms, PermissionLock)
+	}
+	if p&apparmor.MayExecuteMapPermission != 0 {
+		perms = append(perms, PermissionExecuteMap)
+	}
+	if p&apparmor.MayLinkPermission != 0 {
+		perms = append(perms, PermissionLink)
+	}
+	if p&apparmor.MayChangeProfileOnExecPermission != 0 {
+		perms = append(perms, PermissionChangeProfileOnExec)
+	}
+	if p&apparmor.MayChangeProfilePermission != 0 {
+		perms = append(perms, PermissionChangeProfile)
+	}
+	if p&^apparmor.AllFilePermissionMask != 0 {
+		return perms, ErrUnrecognizedFilePermission
+	}
+	return perms, nil
+}
+
+// Returns true if the given permissions list contains the given permission, else false.
+func PermissionsListContains(list []PermissionType, permission PermissionType) bool {
+	for _, perm := range list {
+		if perm == permission {
+			return true
+		}
+	}
+	return false
+}
+
+var allowablePathPatternRegexp = regexp.MustCompile(`^(/|(/[^/*{}]+)*(/\*|(/\*\*)?(/\*\.[^/*{}]+)?)?)$`)
+
+// Checks that the given path pattern is valid.  Returns nil if so, otherwise
+// returns ErrInvalidPathPattern.
+func ValidatePathPattern(pattern string) error {
+	if !allowablePathPatternRegexp.MatchString(pattern) {
+		return ErrInvalidPathPattern
+	}
+	return nil
+}
+
+// Checks that the given outcome is valid.  Returns nil if so, otherwise
+// returns ErrInvalidOutcome.
+func ValidateOutcome(outcome OutcomeType) error {
+	switch outcome {
+	case OutcomeAllow, OutcomeDeny:
+		return nil
+	default:
+		return ErrInvalidOutcome
+	}
+}
+
+// ValidateLifespanParseDuration checks that the given lifespan is valid and
+// that the given duration is valid for that lifespan.  If the lifespan is
+// LifespanTimespan, then duration must be a positive integer representing the
+// number of seconds for which the rule should be valid. Otherwise, it must be
+// 0. Returns an error if any of the above are invalid, otherwise computes the
+// expiration time of the rule based on the current time and the given duration
```

Comment:
There is a lingering reference to access rules, which is not ideal, but the process of validating and parsing a lifespan and duration leads logically to computing the expiration timestamp, which suggests that this is probably fine. If called from accessrules, then the expiration will be used. If called from elsewhere, the caller (perhaps a prompt control panel client) may choose to use or discard the expiration time.